### PR TITLE
Add validation to prevent invalid cipher suites

### PR DESCRIPTION
**Pull request by howardjohn**
Fixes https://github.com/istio/istio/issues/34084

### CODE REVIEW

File: pkg/config/security/security.go [hunk 1]
```diff
@@ -181,3 +182,11 @@ func validateMapKey(key string) error {
 	}
 	return fmt.Errorf("bad key (%s): should have format a[b]", key)
 }
+
+// validCipherSuites contains a list of all ciphers supported in Gateway.server.tls.cipherSuites
+// TODO: DO-NOT-MERGE: extract full supported cipher suite list
+var validCipherSuites = sets.NewSet("ECDHE-ECDSA-AES128-SHA")
```
Comment (lambdai) [inline pkg/config/security/security.go:188]: 1. "ALL" magical words. user should not use though.
2. `bssl ciphers -openssl-name ALL | rg -v PSK`
ECDHE-ECDSA-AES128-GCM-SHA256
ECDHE-RSA-AES128-GCM-SHA256
ECDHE-ECDSA-AES256-GCM-SHA384
ECDHE-RSA-AES256-GCM-SHA384
ECDHE-ECDSA-CHACHA20-POLY1305
ECDHE-RSA-CHACHA20-POLY1305
ECDHE-ECDSA-AES128-SHA
ECDHE-RSA-AES128-SHA
ECDHE-ECDSA-AES256-SHA
ECDHE-RSA-AES256-SHA
AES128-GCM-SHA256
AES256-GCM-SHA384
AES128-SHA
AES256-SHA
DES-CBC3-SHA

File: pkg/config/security/security.go [hunk 2]
```diff
@@ -181,3 +182,11 @@ func validateMapKey(key string) error {
...
+
+func IsValidCipherSuite(cs string) bool {
+	return validCipherSuites.Contains(cs)
```
Comment (lambdai) [inline pkg/config/security/security.go:191]: Can you add that if the cs doesn't start with alphabet, log but let it pass the validation?
The idea is that these list is supposed to be a operator(e.g. exclude blabla) and operand.
I think we the validation is to figure out the obvious error but not block advanced users.

Comment (howardjohn) [reply pkg/config/security/security.go]: Done

File: pkg/config/security/security.go (hunk 1 above)
Comment (howardjohn) [reply pkg/config/security/security.go]: Done ,thanks!

File: pkg/config/security/security.go [hunk 3]
```diff
@@ -181,3 +183,35 @@ func validateMapKey(key string) error {
...
+	"AES128-SHA",
+	"AES256-SHA",
+	"DES-CBC3-SHA",
+)
+
+func IsValidCipherSuite(cs string) bool {
+	if cs == "" || cs == "ALL" {
+		return true
+	}
+	if unicode.IsSymbol(rune(cs[0])) {
```
Comment (lambdai) [inline pkg/config/security/security.go:211]: Great. Thanks!

**Review by lambdai — APPROVED** [review]
LGTM. I have a poor knowledge on openssl. 

This covers all the easy use cases that I am aware of.

**Comment by ramaraochavali** [comment]
Why do not we add this in gateway validation instead of at listener creation? Since this is an invalid config - I think it is OK to add that new validation as any one would have used it like that?

**Comment by howardjohn** [comment]
We add warning at validation because if we make it hard reject, then some
configurations that are previously valid will suddenly stop working when
the  upgrade. Kubernetes API guidance suggests to not make validation
stricter without a new api version

**Comment by ramaraochavali** [comment]
> We add warning at validation because if we make it hard reject, then some
configurations that are previously valid will suddenly stop working when
the  upgrade. 

But this config would have never worked right? Envoy would have rejected it any way so people might not have used it like this. So was thinking it would be good to enforce this. But thinking about this more, when we have mutiple servers in gateway and one is bad, we should not reject all now.

**Comment by ramaraochavali** [comment]
Another idea, how about doing it in 'MergeGateways' and drop that server?

**Comment by howardjohn** [comment]
> Another idea, how about doing it in 'MergeGateways' and drop that server?

Dropping the whole server seems not great, since they may just add a cipher that is invalid and suddenly it breaks? Especially since we only warn about it.

Otherwise the UX right now would be they add a bad cipher, get a warning, and then update the yaml and re-apply. But during that time they have an outage.


> But this config would have never worked right? 

Its actually a bit trickier. The cipher suite rejection logic is not always enacted when we push LDS. If the cert is loaded latter, we may push LDS successfully without any NACK, then later the cert is added and it starts breaking.

**Comment by ramaraochavali** [comment]
> Its actually a bit trickier. The cipher suite rejection logic is not always enacted when we push LDS. If the cert is loaded latter, we may push LDS successfully without any NACK, then later the cert is added and it starts breaking.

Got it. 

My only concern with this is we are doing it every time listener creation happens. But not a huge issue - would have been good if it is done once.

File: pilot/pkg/networking/core/v1alpha3/gateway.go [hunk 4]
```diff
@@ -637,13 +638,26 @@ func buildGatewayListenerTLSContext(
...
 }
 
+// Invalid cipher suites lead Envoy to NACKing. This filters the list down to just the supported set.
+func filteredCipherSuites(suites []string) []string {
+	ret := make([]string, 0, len(suites))
+	for _, s := range suites {
+		if security.IsValidCipherSuite(s) {
+			ret = append(ret, s)
+		} else {
+			log.Debugf("ignoring unsupported cipherSuite: %q", s)
```
Comment (stevenctl) [inline pilot/pkg/networking/core/v1alpha3/gateway.go:655]: nit: including the server name may make it easier to find the offending config

File: pkg/config/security/security.go [hunk 5]
```diff
@@ -181,3 +183,35 @@ func validateMapKey(key string) error {
...
+)
+
+func IsValidCipherSuite(cs string) bool {
+	if cs == "" || cs == "ALL" {
+		return true
+	}
+	if !unicode.IsNumber(rune(cs[0])) && !unicode.IsLetter(rune(cs[0])) {
+		// Not all of these are correct, but this is needed to support advanced cases like - and + operators
+		// without needing to parse the full expression
+		return true
```
Comment (stevenctl) [inline pkg/config/security/security.go:214]: Would `return IsValidCipherSuite(cs[1:])` be valid? Still doesn't catch all cases but it's closer.

**Review by stevenctl — APPROVED** [review]

File: pkg/config/security/security.go [hunk 6]
```diff
@@ -181,3 +183,35 @@ func validateMapKey(key string) error {
...
+	"ECDHE-RSA-CHACHA20-POLY1305",
+	"ECDHE-ECDSA-AES128-SHA",
+	"ECDHE-RSA-AES128-SHA",
+	"ECDHE-ECDSA-AES256-SHA",
+	"ECDHE-RSA-AES256-SHA",
+	"AES128-GCM-SHA256",
+	"AES256-GCM-SHA384",
+	"AES128-SHA",
+	"AES256-SHA",
+	"DES-CBC3-SHA",
```
Comment (nibahl) [inline pkg/config/security/security.go:204]: Hi @howardjohn, we were trying to enable some of the cipher suites in our Istio config but we face 2 issues. 
1. Even after specifying these values in our gateway.yaml file, we do not see some of them when we run the scanner on our endpoint here for TLS 1.2: https://www.ssllabs.com/ssltest/
2. Is there a way to specify the cipher suites for TLS 1.3? We do see some of the cipher suites show up on the above scanner even after not specifying them in our config. Some of them are not allowed as per our company policy. Is there a way to fix this?

Comment (howardjohn) [reply pkg/config/security/security.go:204]: 1. Hard to say without seeing the configdump, logs, etc
2. boringssl doesn't allow custom cipher suites in TLS 1.3. So if you want this in Istio you would need to change boringssl.

**Pull request merged**